### PR TITLE
Stack registry integration

### PR DESCRIPTION
**Pull request by kke**
Implements the commands to interact with the registry:

```
$ kontena stack pull [-f filename.yml] user/stack:version
$ kontena stack push filename.yml
$ kontena stack delete user/stack:version
$ kontena stack search [foo]
$ kontena stack info user/stack
```

And makes stacks loadable / extendable from registry:

```ruby
> r = Kontena::Cli::Stacks::YAML::Reader.new('kke/barstack:1.0.0', from_registry: true)
> puts r.yaml.split("\n").join("\n# ")
# stack: kke/barstack
# version: 1.0.0
# description: Foofoofoo
# services:
#   mysql:
#    extends:
#     stack: kke/dummystack:1.0.0
#     service: foofoo
> r.execute
=> {:stack=>"kke/barstack", :version=>"1.0.0", :name=>"barstack", :expose=>nil, :errors=>[], :notifications=>[], 
:services=>{"mysql"=>{"image"=>"mysql", "environment"=>["FOO=bar"], "secrets"=>[]}
}, :variables=>{}}
```

### CODE REVIEW

**Comment by jnummelin** [comment]
Isn't delete bit confusing as `kontena stack remove` removes the stack from master and `kontena stack delete` from registry? Dunno what it should be, but I'm afraid it will confuse users.

File: cli/lib/kontena/cli/stack_command.rb [hunk 1]
```diff
@@ -23,10 +25,11 @@ class Kontena::Cli::StackCommand < Kontena::Command
   subcommand "logs", "Show logs from stack services", Kontena::Cli::Stacks::LogsCommand
   subcommand "monitor", "Monitor stack services", Kontena::Cli::Stacks::MonitorCommand
   subcommand ["remove","rm"], "Remove a deployed stack", Kontena::Cli::Stacks::RemoveCommand
-  subcommand "push", "Push a stack to stacks registry", Kontena::Cli::Stacks::PushCommand
-  subcommand ["pull", "get"], "Pull a stack from stacks registry", Kontena::Cli::Stacks::PullCommand
-  subcommand "install", "Deploy a stack to Kontena Master", Kontena::Cli::Stacks::InstallCommand
```
Comment (SpComb) [inline cli/lib/kontena/cli/stack_command.rb]: What happened to `kontena stack install`?

File: cli/lib/kontena/cli/stacks/search_command.rb [hunk 2]
```diff
@@ -5,12 +5,14 @@ class SearchCommand < Kontena::Command
...
-
-    requires_current_account_token
+    parameter "[QUERY]", "Query string"
 
     def execute
-      puts stacks_client.search(query).inspect
+      results = stacks_client.search(query.to_s)
+      exit_with_error 'Nothing found' if results.empty?
+      results.each do |stack|
+        puts stack['name']
```
Comment (SpComb) [inline cli/lib/kontena/cli/stacks/search_command.rb]: Is this supposed to be `name`?

Comment (kke) [inline cli/lib/kontena/cli/stacks/search_command.rb]: Well, yes and no.

It would make sense that it would say 'stack'.  Currently it returns 'name'.

**Comment by kke** [comment]
Delete prompts say "about to blahblah from registry", I suppose it helps with the confusion.

Also the command descriptions in --help say "from the stacks registry"

**Comment by kke** [comment]
It's still a concern though. Making a stack version that sends your stuff
somewhere is easy.

25.11.2016 17.34 "Tero Marttila" <notifications@github.com> kirjoitti:

*@SpComb* commented on this pull request.
------------------------------

In cli/lib/kontena/cli/stacks/yaml/reader.rb
<https://github.com/kontena/kontena/pull/1403#pullrequestreview-10178299>:

> @@ -17,7 +17,15 @@ def initialize(file, skip_validation: false, skip_variables: false, replace_miss
         require_relative 'opto/prompt_resolver'

         @file = file
-        @raw_content = File.read(File.expand_path(file))
+        @from_registry = from_registry
+
+        if from_registry?
+          require 'shellwords'
+          @raw_content = Kontena.run("stack pull --return
#{file.shellescape}", returning: :result)

Are these shellescape'd stack file names evaluated from remotely pulled
stacks? Does this shellescape properly handle something bizarre cases
like extends:
stack: -f ~/.ssh/authorized_keys eve/maliciousstack, which might cause the
kontena-cli to replace the user's ~/.ssh/authorized_keys file?

Should this just use a function call instead of running shell commands?

—
You are receiving this because you were assigned.
Reply to this email directly, view it on GitHub
<https://github.com/kontena/kontena/pull/1403#pullrequestreview-10178299>,
or mute the thread

**Comment by SpComb** [comment]
The `kontena stack info terom/wordpress` seems to be broken for the new `[ {version:"..."} ]` API response:

```
Running #<Kontena::MainCommand:0x0000000246b7f0> -- callback matcher = 'nil'
Running #<Kontena::Cli::StackCommand:0x00000002468f78> -- callback matcher = 'nil'
Running #<Kontena::Cli::Stacks::InfoCommand:0x0000000246fc10> -- callback matcher = 'stacks info'
Running Kontena.run("stack pull --return terom/wordpress", returning: result
Running #<Kontena::MainCommand:0x0000000270a1c8> -- callback matcher = 'nil'
Running #<Kontena::Cli::StackCommand:0x000000027089e0> -- callback matcher = 'nil'
Running #<Kontena::Cli::Stacks::PullCommand:0x000000027135c0> -- callback matcher = 'stacks pull'
D, [2016-11-27T13:10:30.643428 #27074] DEBUG -- CONFIG: Loading configuration from /home/kontena/.kontena_client.json
D, [2016-11-27T13:10:30.643754 #27074] DEBUG -- CONFIG: Configuration loaded with 3 servers.
D, [2016-11-27T13:10:30.643776 #27074] DEBUG -- CONFIG: Current master: development
D, [2016-11-27T13:10:30.643809 #27074] DEBUG -- CONFIG: Current grid: development
D, [2016-11-27T13:10:30.643912 #27074] DEBUG -- CLIENT: Excon opts: {:omit_default_port=>true, :connect_timeout=>5, :read_timeout=>30, :write_timeout=>5, :ssl_verify_peer=>true}
Command completed, result: "stack: terom/wordpress\nversion: 0.2.0\nservices:\n  wordpress:\n    image: wordpress:4.1\n    stateful: true\n    ports:\n      - 80:80\n    environment:\n      - WORDPRESS_DB_HOST=mysql\n      - WORDPRESS_DB_USER=root\n      - WORDPRESS_DB_PASSWORD=secret\n  mysql:\n    image: mariadb:5.5\n    stateful: true\n    environment:\n      - MYSQL_ROOT_PASSWORD=secret\n" status: 0
CURRENT VERSION
-----------------
* Version: 0.2.0

AVAILABLE VERSIONS
-------------------
D, [2016-11-27T13:10:31.572657 #27074] DEBUG -- CLIENT: Excon opts: {:omit_default_port=>true, :connect_timeout=>5, :read_timeout=>30, :write_timeout=>5, :ssl_verify_peer=>true}
bundler: failed to load command: bin/kontena (bin/kontena)
ArgumentError: comparison of Hash with Hash failed
  /home/kontena/kontena/kontena/cli/lib/kontena/cli/stacks/info_command.rb:22:in `sort'
  /home/kontena/kontena/kontena/cli/lib/kontena/cli/stacks/info_command.rb:22:in `execute'
  /home/kontena/kontena/kontena/cli/lib/kontena/command.rb:181:in `run'
  /home/kontena/kontena/kontena/cli/vendor/bundle/ruby/2.3.0/gems/clamp-1.1.1/lib/clamp/subcommand/execution.rb:11:in `execute'
  /home/kontena/kontena/kontena/cli/lib/kontena/command.rb:181:in `run'
  /home/kontena/kontena/kontena/cli/vendor/bundle/ruby/2.3.0/gems/clamp-1.1.1/lib/clamp/subcommand/execution.rb:11:in `execute'
  /home/kontena/kontena/kontena/cli/lib/kontena/command.rb:181:in `run'
  /home/kontena/kontena/kontena/cli/vendor/bundle/ruby/2.3.0/gems/clamp-1.1.1/lib/clamp/command.rb:132:in `run'
  bin/kontena:16:in `<top (required)>'
```

**Comment by kke** [comment]
Stack info command now works with the new output from registry

**Comment by jnummelin** [comment]
what about the `delete` vs. `push -d` for removing stack in the registry?

**Comment by kke** [comment]
Delete is now `push -d`

**Comment by SpComb** [comment]
The `kontena stack push -d` UX is indeed confusing, but not sure what to do about it

```
$ bundle exec bin/kontena stack push -d ~/kontena/kontena/docs/examples/wordpress/kontena.yml 
About to delete an entire stack and all of its versions from the registry
Destructive command. To proceed, type "/home/kontena/kontena/kontena/docs/examples/wordpress/kontena.yml" or re-run this command with --force option.
> Enter '/home/kontena/kontena/kontena/docs/examples/wordpress/kontena.yml' to confirm:  
$ bundle exec bin/kontena stack push -d --help
About to delete an entire stack and all of its versions from the registry
Destructive command. To proceed, type "--help" or re-run this command with --force option.
> Enter '--help' to confirm:      
```

**Comment by kke** [comment]
Maybe at minimum -d should try to guess a bit / validate that the param
looks like a stack name.

28.11.2016 6.47 ip. "Tero Marttila" <notifications@github.com> kirjoitti:

File: cli/lib/kontena/cli/stack_command.rb [hunk 3]
```diff
@@ -24,11 +20,8 @@ class Kontena::Cli::StackCommand < Kontena::Command
   subcommand "logs", "Show logs from services in a stack", Kontena::Cli::Stacks::LogsCommand
   subcommand "monitor", "Monitor services in a stack", Kontena::Cli::Stacks::MonitorCommand
   subcommand "build", "Build images listed in a stack file and push them to an image registry", Kontena::Cli::Stacks::BuildCommand
-  subcommand "push", "Push (or delete) a stack into the stacks registry", Kontena::Cli::Stacks::PushCommand
-  subcommand "pull", "Pull a stack from the stacks registry", Kontena::Cli::Stacks::PullCommand
-  subcommand "search", "Search for stacks in the stacks registry", Kontena::Cli::Stacks::SearchCommand
-  subcommand "info", "Show info about a stack in the stacks registry", Kontena::Cli::Stacks::InfoCommand
-  
+  subcommand "registry", "Stack registry related commands", Kontena::Cli::Stacks::RegistryCommand
```
Comment (kke) [inline cli/lib/kontena/cli/stack_command.rb]: maybe add "reg" or something for shorthand? (or merge the thing that allows typing just enough to make subcommand  unique)

File: cli/lib/kontena/cli/stacks/registry/remove_command.rb [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+module Kontena::Cli::Stacks::Registry
+  class RemoveCommand < Kontena::Command
+    include Kontena::Cli::Common
+    include Kontena::Cli::Stacks::Common
+    include Kontena::Cli::Stacks::Common::StackNameParam
+
+    banner "Removes a stack (or version) from the stack registry. Use user/stack_name or user/stack_name:version."
+
+    parameter "[FILENAME]", "Stack file path"
```
Comment (SpComb) [inline cli/lib/kontena/cli/stacks/registry/remove_command.rb]: Remove the `[FILENAME]`?

```
Usage:
    kontena stack registry remove [OPTIONS] STACK_NAME [FILENAME]

  Removes a stack (or version) from the stack registry. Use user/stack_name or user/stack_name:version.

Parameters:
    STACK_NAME                    Stack name, for example user/stackname or user/stackname:version
    [FILENAME]                    Stack file path

Options:
    -f, --force                   Force delete
    -h, --help                    print help
```

File: cli/lib/kontena/cli/stacks/registry/remove_command.rb [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+
+    option ['-f', '--force'], :flag, "Force delete"
+
+    requires_current_account_token
+
+    def execute
+      unless force?
+        if stack_name.include?(':')
```
Comment (SpComb) [inline cli/lib/kontena/cli/stacks/registry/remove_command.rb]: This doesn't work:

```
kontena stack registry remove terom/wordpress:0.3.2
About to delete an entire stack and all of its versions from the stacks registry
Destructive command. To proceed, type "terom/wordpress" or re-run this command with --force option.
> Enter 'terom/wordpress' to confirm: 
```

Probably needs to be `unless stack_version.nil?`

File: cli/lib/kontena/cli/stacks/registry/remove_command.rb [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+          puts "About to delete #{pastel.cyan(stack_name)} from the stacks registry"
+          confirm
+        else
+          puts "About to delete an entire stack and all of its versions from the stacks registry"
+          confirm_command(stack_name)
+        end
+      end
+      spinner "Removing #{pastel.cyan(stack_name)} from the registry" do
+        stacks_client.destroy(stack_name)
```
Comment (SpComb) [inline cli/lib/kontena/cli/stacks/registry/remove_command.rb]: `.destroy(stack_name, stack_version)`

File: cli/lib/kontena/cli/stacks/registry/remove_command.rb (hunk 5 above)
Comment (kke) [inline cli/lib/kontena/cli/stacks/registry/remove_command.rb]: yes, it was like that when it was independent previously.

File: cli/lib/kontena/cli/stacks/registry/remove_command.rb (hunk 6 above)
Comment (kke) [inline cli/lib/kontena/cli/stacks/registry/remove_command.rb]: (y)

File: cli/lib/kontena/cli/stacks/registry/show_command.rb [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+
+module Kontena::Cli::Stacks::Registry
+  class ShowCommand < Kontena::Command
+    include Kontena::Cli::Common
+    include Kontena::Cli::Stacks::Common
+    include Kontena::Cli::Stacks::Common::StackNameParam
+
+    banner "Shows information about a stack on the stacks registry"
+
+    option ['-v', '--versions'], :flag, "Only list available versions"
```
Comment (SpComb) [inline cli/lib/kontena/cli/stacks/registry/show_command.rb]: Not implemented

Comment (kke) [inline cli/lib/kontena/cli/stacks/registry/show_command.rb]: There may be a push missing or it got lost somewhere, I'm sure I implemented it :)

File: cli/lib/kontena/cli/stacks/registry/show_command.rb [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+
+    requires_current_account_token
+
+    def execute
+      stack = ::YAML.load(stacks_client.show(stack_name))
```
Comment (kke) [inline cli/lib/kontena/cli/stacks/registry/show_command.rb]: As this one too.  Using `user/stack:version` showed info from that version and just `user/stack` would show info on the 'latest'.

Will push in the morning from office.

File: cli/lib/kontena/cli/stacks/registry/show_command.rb (hunk 7 above)
Comment (jnummelin) [inline cli/lib/kontena/cli/stacks/registry/show_command.rb]: Should we leave this now as is, and really implement it in another PR? I'd really like to get this merged :)

Comment (kke) [inline cli/lib/kontena/cli/stacks/registry/show_command.rb]: It's just a case of adding an if. I can try, I'm on windows and don't have my terminal magic.

Comment (jnummelin) [inline cli/lib/kontena/cli/stacks/registry/show_command.rb]: Ok. Give me a ping if you can or cannot.

Comment (SpComb) [inline cli/lib/kontena/cli/stacks/registry/show_command.rb]: > There may be a push missing or it got lost somewhere, I'm sure I implemented it :)

It got lost in @jakolehm's refactoring

Comment (kke) [inline cli/lib/kontena/cli/stacks/registry/show_command.rb]: Hmm wait a second.

stacks_client.show returns json, not yaml?

File: cli/lib/kontena/cli/stacks/registry/show_command.rb [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+      #p stack
+      puts "#{stack['stack']}:"
+      puts "  latest_version: #{stack['version']}"
+      puts "  expose: #{stack['expose'] || '-'}"
+      puts "  description: #{stack['description'] || '-'}"
+
+      puts "  available_versions:"
+      stacks_client.versions(stack_name).map { |s| s['version']}.sort.reverse_each do |version|
```
Comment (kke) [inline cli/lib/kontena/cli/stacks/registry/show_command.rb]: It could use `sort {|a,b| Gem::Version.new(a) <=> Gem::Version.new(b)}`

Comment (kke) [inline cli/lib/kontena/cli/stacks/registry/show_command.rb]: or replace the .map with `.map {|v| Gem::Version.new(v['version']}.sort`

File: cli/lib/kontena/cli/stacks/registry/show_command.rb (hunk 7 above)
Comment (SpComb) [inline cli/lib/kontena/cli/stacks/registry/show_command.rb]: `GET /stack/terom/redis` returns `Content-Type: text/x-yaml; charset=utf-8` and `Client#parse_response` returns the response body as a String.

Comment (jakolehm) [inline cli/lib/kontena/cli/stacks/registry/show_command.rb]: Sorry about the hassle, had to make quick changes to ux before it's too late.

Comment (kke) [inline cli/lib/kontena/cli/stacks/registry/show_command.rb]: `.show` sends `Accept: application/json` header

**Comment by SpComb** [comment]
`kontena stack build -f ~/kontena/examples/whoami/kontena.yml`

```
NameError: undefined local variable or method `service_prefix' for #<Kontena::Cli::Stacks::BuildCommand:0x000000027c20e8>
Did you mean?  service_list
  /home/kontena/kontena/kontena/cli/lib/kontena/cli/stacks/build_command.rb:18:in `execute'
  /home/kontena/kontena/kontena/cli/lib/kontena/command.rb:187:in `run'
  /home/kontena/kontena/kontena/cli/vendor/bundle/ruby/2.3.0/gems/clamp-1.1.1/lib/clamp/subcommand/execution.rb:11:in `execute'
  /home/kontena/kontena/kontena/cli/lib/kontena/command.rb:187:in `run'
  /home/kontena/kontena/kontena/cli/vendor/bundle/ruby/2.3.0/gems/clamp-1.1.1/lib/clamp/subcommand/execution.rb:11:in `execute'
  /home/kontena/kontena/kontena/cli/lib/kontena/command.rb:187:in `run'
  /home/kontena/kontena/kontena/cli/vendor/bundle/ruby/2.3.0/gems/clamp-1.1.1/lib/clamp/command.rb:132:in `run'
  bin/kontena:16:in `<top (required)>'
```

**Comment by jnummelin** [comment]
I'll merge this now and let's continue the fixes on separate PR's

**Pull request merged**